Implemented mailing in auth controller

SedovMykyta opened the pull request:
feat - Implemented mailing in auth controller*;
feat - Created email service;
fix - Fixed names in claim parser.

Commit:
feat - Implemented mailing in auth service;
feat - Created email service;
fix - Fixed names in claim parser.

Vladislawv commented on MyBlog.API/Controllers/AuthController.cs:
```diff
@@ -12,10 +13,11 @@ namespace MyBlog.Controllers;
     private readonly IAuthService _authService;
-    
-    public AuthController(IAuthService authService)
+    private readonly IEmailService _emailService;
+    public AuthController(IAuthService authService, IEmailService emailService)
```
отступ

Vladislawv commented on MyBlog.API/Controllers/AuthController.cs:
```diff
@@ -12,10 +13,11 @@ namespace MyBlog.Controllers;
+    public AuthController(IAuthService authService, IEmailService emailService)
     {
         _authService = authService;
+        _emailService = emailService;
     }
```
mailingService не лучше ?

Vladislawv commented on MyBlog.API/Controllers/AuthController.cs:
```diff
@@ -32,6 +34,8 @@ public async Task<IActionResult> RegisterAsync([FromForm] UserDtoInput userInput
     {
         await _authService.RegisterAsync(userInput);
 
+        await _emailService.SendMessageAsync("You success register on site 'My Blog'.", userInput.Email);
+
```
successfull

Vladislawv commented on MyBlog.Services/Areas/Mailing/EmailService.cs:
```diff
@@ -0,0 +1,38 @@
+
+    public async Task SendMessageAsync(string message, string recipientEmail)
+    {
+        var mimeMessage = new MimeMessage();
+        mimeMessage.From.Add(new MailboxAddress("My blog", _email));
+        mimeMessage.To.Add(new MailboxAddress("",recipientEmail));
+        mimeMessage.Subject = message;
+        mimeMessage.Body = new TextPart(MimeKit.Text.TextFormat.Html)
```
сделай отправку сообщений анонимной, что бы получатели не видели, кому ещё приходили письма

Vladislawv commented on MyBlog.Services/Areas/Mailing/EmailService.cs:
```diff
@@ -0,0 +1,38 @@
+            await client.DisconnectAsync(true);
+        }
+    } 
+}
```
так же, можно реализовать метод рассылки от одного ко многим (админ - юзеры)

Commit:
fix - Fixed code review remarks;
feat - Added new user subscriptions entity;
feat - Created new migrations;
feat - Created mailing service\controller;
feat - Added mail settings class;
feat - Implemented mailing in auth service\controller;
fix - Deleted whitespace in dto`s;
fix - Deleted unless virtual in navigation property in all entity;

Vladislawv commented on MyBlog.API/Controllers/AuthController.cs:
```diff
@@ -12,25 +13,30 @@ namespace MyBlog.Controllers;
     [ProducesResponseType(StatusCodes.Status200OK)]
     [ProducesResponseType(StatusCodes.Status400BadRequest)]
-    public async Task<IActionResult> RegisterAsync([FromForm] UserDtoInput userInput)
+    public async Task<IActionResult> RegisterAsync([FromForm] UserDtoInput userInput, [FromForm] bool isSubscribeToEmail)
```
бульку можно FromQuery

Vladislawv commented on MyBlog.API/Controllers/AuthController.cs:
```diff
@@ -12,25 +13,30 @@ namespace MyBlog.Controllers;
-        await _authService.RegisterAsync(userInput);
+        await _authService.RegisterAsync(userInput, isSubscribeToEmail);
+
+        await _mailingService.SendEmailToUserAsync("You successfully register on site 'My Blog'.", userInput.Email);
```
registered

Vladislawv commented on MyBlog.API/Controllers/MailingController.cs:
```diff
@@ -0,0 +1,79 @@
+
+namespace MyBlog.Controllers;
+
+[Route("api/article")]
+[ApiController]
```
nice root, man

SedovMykyta replied:
ахахах, оце я прошляпился

Vladislawv commented on MyBlog.API/Controllers/MailingController.cs:
```diff
@@ -0,0 +1,79 @@
+    {
+        _mailingService = mailingService;
+        _currentUserJwtInfo = Task.Run(() => parse.ToJwtInfo(HttpContext.User.Identity as ClaimsIdentity));
+
+    }
```
?

Vladislawv commented on MyBlog.API/Controllers/MailingController.cs:
```diff
@@ -0,0 +1,79 @@
+    [Authorize(Roles = "Admin")]
+    [ProducesResponseType(StatusCodes.Status200OK)]
+    [ProducesResponseType(StatusCodes.Status404NotFound)]
+    public async Task<IActionResult> SendEmailToUserAsync([FromForm] string message, [FromForm] string recipientEmail)
```
логичнее в начале указать получаетеля, а потом сообщение
посмотри везде

Vladislawv commented on MyBlog.API/Controllers/MailingController.cs:
```diff
@@ -0,0 +1,79 @@
+    /// <returns>Returns Ok</returns>
+    /// <response code="200">Success</response>
+    /// <response code="400">UserNotFound</response>
+    [HttpPost ("changeSubscribe")]
+    [Authorize]
+    [ProducesResponseType(StatusCodes.Status200OK)]
+    [ProducesResponseType(StatusCodes.Status400BadRequest)]
+    public async Task ChangeSubscribeOnMailingAsync([FromForm] bool isSubscribeToEmail)
+    {
+        await _mailingService.ChangeSubscribeOnMailingAsync(isSubscribeToEmail, _currentUserJwtInfo.Result.Id);
+    }
```
лучше разбей на 2 ендпоинта

Vladislawv commented on MyBlog.API/appsettings.json:
```diff
@@ -13,5 +13,13 @@
     "Audience": "User"
   },
+  "MailSettings": {
+    "DisplayName": "My blog",
+    "From": "Stores in UserSecrets",
+    "Host": "smtp.gmail.com",
+    "Password": "Stores in UserSecrets",
+    "Port": 465,
+    "UseSSL": true
```
лучше всё в сикреты

Vladislawv commented on MyBlog.Infrastructure/Entities/User.cs:
```diff
@@ -16,5 +16,6 @@ public class User
     public IList<Like> Likes { get; set; }
-    public IList<Dislike> Dislikes { get; set; } 
+    public IList<Dislike> Dislikes { get; set; }
+    public UserSubscription Subscription { get; set; }
 }
```
я бы, всё же, заменил на бульку

Vladislawv commented on MyBlog.Services/Areas/Mailing/MailingService.cs:
```diff
@@ -0,0 +1,97 @@
+        _settings = settings.Value;
+    }
+
+    public async Task SendEmailToUserAsync(string message, string recipientEmail)
+    {
+        await ThrowIfEmailNotFound(recipientEmail);
+        
+        var mimeMessage = new MimeMessage();
+        mimeMessage.From.Add(new MailboxAddress(_settings.DisplayName, _settings.From));
+        mimeMessage.To.Add(new MailboxAddress("",recipientEmail));
+        mimeMessage.Subject = message;
+        mimeMessage.Body = new TextPart(MimeKit.Text.TextFormat.Html)
+        {
+            Text = message
+        };
+
+        using (var client = new SmtpClient()) 
+        {
+            await client.ConnectAsync(_settings.Host, _settings.Port, _settings.UseSSL);
+            await client.AuthenticateAsync(_settings.From, _settings.Password );
+            await client.SendAsync(mimeMessage);
+
+            await client.DisconnectAsync(true);
+        }
+    }
+
+    public async Task SendEmailToSubscribedUsersAsync(string message)
+    {
+        var recipientEmails = await GetAllUsersEmailAsync();
+
+        foreach (var email in recipientEmails)
+        {
+            await SendEmailToUserAsync(message, email);
+        }
+    }
```
лучше поменять местами методы

Vladislawv commented on MyBlog.Services/Areas/Mailing/MailingService.cs:
```diff
@@ -0,0 +1,97 @@
+        var emailsSubscribedUsers = await _context.Users.
+            Where(user => user.Role == Role.User).
+            Include(user => user.Subscription).
+            Where(user => user.Subscription.IsSubscribedToEmail == true).
+            Select(user => user.Email).
```
можно ли убрать == true ?

Vladislawv commented on MyBlog.Services/Areas/Mailing/MailingService.cs:
```diff
@@ -0,0 +1,97 @@
+    }
+
+
+    private async Task<UserSubscription> GetUserSubscriptionByUserIdAsync(int id)
+    {
+        var user = await _context.UserSubscriptions.FirstOrDefaultAsync(user => user.UserId == id)
+                   ?? throw new NotFoundException($"User with Id: {id} is not found");
+
+        return user;
+    }
```
метод по сути должен вызываться в кейсе при подписке/отписке от рассылки => можно не выносить в отдельный метод

Commit:
fix - Fixed code review remarks.

Vladislawv commented on MyBlog.API/Controllers/MailingController.cs:
```diff
@@ -62,18 +61,32 @@ public async Task<IActionResult> SendEmailToSubscribedUsersAsync([FromForm] stri
+    public async Task UnsubscribeOnMailingAsync()
     {
-        await _mailingService.ChangeSubscribeOnMailingAsync(isSubscribeToEmail, _currentUserJwtInfo.Result.Id);
+        await _mailingService.UnsubscribeOnMailingAsync(_currentUserJwtInfo.Result.Id);
```
как называется твой контроллер ?

Vladislawv commented on MyBlog.API/Controllers/MailingController.cs:
```diff
@@ -62,18 +61,32 @@ public async Task<IActionResult> SendEmailToSubscribedUsersAsync([FromForm] stri
+    /// <response code="200">Success</response>
+    /// <response code="400">UserNotFound</response>
+    [HttpPost ("changeSubscribe")]
+    [Authorize]
```
не-вни-ма-тель-ный

Vladislawv commented on MyBlog.Services/Areas/Mailing/IMailingService.cs:
```diff
@@ -2,9 +2,11 @@ namespace MyBlog.Service.Areas.Mailing;
-
-    public Task ChangeSubscribeOnMailingAsync(bool isSubscribeToEmail, int id);
+    
+    public Task SubscribeOnMailingAsync(int id);
+    
+    public Task UnsubscribeOnMailingAsync(int id);
```
не стоит дублировать название сервиса
Subscribe
Unsubscribe

Vladislawv commented on MyBlog.Services/Areas/Mailing/MailingService.cs:
```diff
@@ -87,7 +91,7 @@ private async Task<List<string>> GetAllUsersEmailAsync()
             Include(user => user.Subscription).
-            Where(user => user.Subscription.IsSubscribedToEmail == true).
+            Where(user => user.Subscription.IsSubscribedToEmail).
             Select(user => user.Email).
```
IsSubscribed на первое место в linq

Vladislawv commented on MyBlog.Services/Areas/Mailing/MailingService.cs:
```diff
@@ -87,7 +91,7 @@ private async Task<List<string>> GetAllUsersEmailAsync()
         var emailsSubscribedUsers = await _context.Users.
             Where(user => user.Role == Role.User).
```
название метода думай сам)

Commit:
fix - Fixed code review remarks.

Vladislawv commented on MyBlog.Services/Areas/Mailing/MailingService.cs:
```diff
@@ -86,12 +85,12 @@ private async Task ThrowIfEmailNotFound(string email)
 
-    private async Task<List<string>> GetAllUsersEmailAsync()
+    private async Task<List<string>> GetEmailsSubscribedUsersAsync()
     {
```
GetSubscribedEmailsAsync pretty, yeah ?

Commit:
fix - Fixed code review remarks;
fix - Fixed namespace in filters.
